local/http2_handler.go

```diff
@@ -0,0 +1,293 @@
+		}
+	}
+}
+
+func (s *HTTP2Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	s.logger.Info("We got a local request!")
+	r.URL.Host = s.LocalEndpoint
+	// TODO: Figure out support for https
+	r.URL.Scheme = "http"
+	r.Host = s.LocalEndpoint
+	r.RequestURI = ""
+
+	// We ignore the error ONLY because it will be forwarded
+	// to the end user
+	// TODO: Handle this error
+	resp, err := s.fClient.Do(r)
+	if err != nil {
+		s.logger.Error(err)
+	}
+
+	// Delete this so we don't copy it over
+	// Will be handled by http.ResponseWriter
+	resp.Header.Del("Content-Length")
+
+	for key, values := range resp.Header {
+		for _, value := range values {
+			w.Header().Add(key, value)
+		}
+	}
+
+	w.WriteHeader(resp.StatusCode)
+
+	defer resp.Body.Close()
+
+	nr, err := io.Copy(w, resp.Body)
+	if err != nil {
+		s.logger.Errorf("Could not copy response body")
+		return
+	}
+	s.logger.Infof("Copied %d bytes between connection bodies", nr)
+}
+
+// Close closes the listener and TCP connection
+func (s *HTTP2Handler) Close() error {
+	err := s.control.Close()
+	if err != nil {
+		s.logger.Errorf("Control TCP conn close: %s", err)
+	}
+	for _, c := range s.conns {
+		err = c.Close()
+		if err != nil {
+			s.logger.Errorf("Proxy http2 conn close: %s", err)
+		}
+	}
+	return err
+}
+
+// dial opens an unencrypted TCP connection to a server
+func (s *HTTP2Handler) dial() (*net.TCPConn, error) {
+	conn, err := net.Dial("tcp", s.RemoteEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	s.logger.Info("Established TCP connection")
+	tcpConn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return nil, errors.New("Error: could not cast tcp connection")
+	}
+
+	return tcpConn, nil
+}
+
+func (s *HTTP2Handler) dialControl() (net.Conn, error) {
+	conn, err := s.dial()
+	if err != nil {
+		return nil, err
+	}
+
+	cConn, err := s.genericTLSWrap(conn)
+	if err != nil {
+		return nil, err
+	}
+
+	s.logger.Info("TCP Wrapped with TLS")
+
+	return cConn, nil
+}
+
+func (s *HTTP2Handler) genericTLSWrap(conn *net.TCPConn) (*tls.Conn, error) {
+	return wnet.GenericTLSWrap(conn, s.tlsConfig, tls.Client)
+}
+
+// This wrapper fulfills the requirement for specifying the 'h2' ALPN TLS negotiation for
+// TLS enabled http2 connections
+//
+// NOTE: The ALPN is a requirement of the spec for HTTP/2 capability discovery
+// While technically the golang implementation will allow us not to perform ALPN,
+// this breaks the http/2 spec. The goal here is to follow the RFC to the letter
+// as documented in http://httpwg.org/specs/rfc7540.html#starting
+func (s *HTTP2Handler) http2ALPNTLSWrap(conn *net.TCPConn) (*tls.Conn, error) {
+	return wnet.HTTP2ALPNTLSWrap(conn, s.tlsConfig, tls.Client)
+}
+
+func (s *HTTP2Handler) heartbeat() {
+	// set lastPing to something sane
+	lastPing := time.Unix(atomic.LoadInt64(&s.lastPongAt)-1, 0)
+	ping := time.NewTicker(pingInterval)
+	pongCheck := time.NewTicker(time.Second)
+
+	defer func() {
+		s.control.Close()
```

Comment:
Hmm.  I'm not sure on this right now. Technically a missed heartbeat means there's an issue with the control conn - and we assume that the tunnels/session are compromised in that case. I think we have 2 options. Try a total reset and attempt to get into a healthy state again. Or... Hard fail on purpose. I'm personally a proponent of a hard fail. This way we don't get into a spin-loop of trying to fix a problem that may not be fixable. Better IMO to hard fail - and have the system the agent is running on have an arbiter over it. This way as well - an arbiter can easily report fatal issues rather than WH not alerting to an issue